File: output/gemma4-e4b/avatar/Java/Python/atcoder_ABC122_D.py
import sys

class atcoder_ABC122_D:
    TOKENS = ["A", "C", "G", "T"]
    MOD = 1000000007

    def __init__(self):
        # Read N from standard input
        try:
            # Read the single integer N
            N = int(sys.stdin.read().strip())
        except:
            # Handle case where input might be empty or malformed
            N = 0
        
        self.N = N
        # Initialize memoization table: list of dictionaries
        self.memo = [{} for _ in range(N + 1)]

    def calc(self):
        # Start DFS from current=0, last3="TTT"
        return self.dfs(0, "TTT")

    def is_ok(self, last4: str) -> bool:
        # Constraint 1: Check if last4 contains "AGC"
        if "AGC" in last4:
            return False
        
        # Constraint 2: Check swaps
        # We must replicate the sequential modification of the character array from Java
        
        # Start with a mutable list representation of last4
        vals = list(last4)
        
        # The loop runs for i = 0, 1, 2 (since last4 has length 4)
        for i in range(3):
            # Swap logic: vals[i] <-> vals[i+1]
            # Note: The Java code uses the original last4 characters for the swap values, 
            # but applies them to the current state of 'vals'.
            
            # Java: vals[i] = last4.charAt(i + 1);
            # Java: vals[i + 1] = last4.charAt(i);
            
            # To replicate the Java behavior exactly (where the swap uses the original

File: output/gemma4-e4b/avatar/Python/Java/atcoder_ABC122_D.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class atcoder_ABC122_D {

    private static int N;
    private static final long MOD = 1_000_000_007L;
    // memo[cur] maps last3 (String) -> count (Long)
    private static Map<String, Long>[] memo;

    /**
     * Checks if the 4-character string 'last4' is "okay" based on the specific swapping rule.
     * It checks the original string and the strings resulting from swapping (i, i-1) for i=1, 2, 3.
     */
    private static boolean ok(String last4) {
        char[] originalChars = last4.toCharArray();

        for (int i = 0; i < 4; i++) {
            // Create a fresh copy of the characters for each iteration
            char[] t = originalChars.clone();

            // Perform the swap if i >= 1
            if (i >= 1) {
                // Swap t[i] and t[i-1]
                char temp = t[i];
                t[i] = t[i - 1];
                t[i - 1] = temp;
            }

            String currentString = new String(t);

            // Check for "AGC"
            if (currentString.contains